src/metabase/upload.clj

```diff
@@ -557,25 +561,38 @@
           _                  (check-schema (dissoc normed-name->field auto-pk-column-name) header)
           col-upload-types   (map (comp base-type->upload-type :base_type normed-name->field) normed-header)
           parsed-rows        (parse-rows col-upload-types rows)]
-      (try
-        (when create-auto-pk?
-          (driver/add-columns! driver
-                               (:id database)
-                               (table-identifier table)
-                               {(keyword auto-pk-column-name) (driver/upload-type->database-type driver ::auto-incrementing-int-pk)}))
-        (driver/insert-into! driver
+      (when create-auto-pk?
+        (driver/add-columns! driver
                              (:id database)
                              (table-identifier table)
-                             normed-header
-                             parsed-rows)
-        (scan-and-sync-table! database table)
-        (when create-auto-pk?
-          (let [auto-pk-field (table-id->auto-pk-column (:id table))]
-            (t2/update! :model/Field (:id auto-pk-field) {:display_name (:name auto-pk-field)})))
-        {:row-count (count parsed-rows)}
-        (catch Throwable e
-          ;; TODO: insert rows failure handling
-          (throw (ex-info (ex-message e) {:status-code 400})))))))
+                             {(keyword auto-pk-column-name) (driver/upload-type->database-type driver ::auto-incrementing-int-pk)}))
+      ;; If the insert fails, we need to delete the rows we just inserted. So we:
+      ;; 1. get the max PK value before inserting
+      ;; 2. insert the rows
+      ;; 3. if insertion fails, delete the rows with PKs greater than the max PK value
```

Comment:
I'm hoping I can replace all this with database transactions later, but I'm still trying to figure out why transactions aren't rolling back correctly with MySQL and H2, and this works fine for now.